### ISSUE TITLE
Plone 5.2 / Python 3 compatibility

### ISSUE THREAD

**Issue by tisto** [collective/collective.mailchimp#43]
We need to:

- [x] Fix the Plone 5.2 / Python 2.7 Build: https://travis-ci.org/collective/collective.mailchimp/jobs/507933819
- [ ] Make a Plone 5.2 compatible release
- [ ] Fix the Plone 5.2 / Python 3.6 Build
- [ ] Fix the Plone 5.2 / Python 3.7 Build
- [ ] Make a Python 3 compatible release

**Comment by tisto** [comment]
@erral I made a todo list for a Plone 5.2 / Python 3 compatible release. I fixed the Travis build and made sure it runs the actual code. Problem right now are two failing Plone 5.2 tests. First one looks like the forms changed on 5.2, second one complains about css resource registration:

```
Failure in test test_form_with_email_address (collective.mailchimp.tests.test_newsletter.TestUnsubscribeNewsletterView)
Traceback (most recent call last):
  File "/opt/python/2.7.15/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/travis/build/collective/collective.mailchimp/src/collective/mailchimp/tests/test_newsletter.py", line 161, in test_form_with_email_address
    'data': {u'status': u'unsubscribed'},
  File "/opt/python/2.7.15/lib/python2.7/unittest/case.py", line 513, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/2.7.15/lib/python2.7/unittest/case.py", line 836, in assertDictEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/opt/python/2.7.15/lib/python2.7/unittest/case.py", line 410, in fail
    raise self.failureException(msg)
AssertionError: {'endpoint': u'lists/f6257645gs/members/06f12badc3b5fffc57576822131ded7c', 'data [truncated]... != {'endpoint': u'lists/f6257645gs/members/06f12badc3b5fffc57576822131ded7c', 'data [truncated]...
- {'data': {},
+ {'data': {u'status': u'unsubscribed'},
   'endpoint': u'lists/f6257645gs/members/06f12badc3b5fffc57576822131ded7c'}
    20/30 (66.7%)/home/travis/build/collective/collective.mailchimp/src/collective/mailchimp/tests/test_portlet.py:40: PendingDeprecationWarning: Please use assertTrue instead.
  self.failUnless(IPortletAssignment.providedBy(portlet))
/home/travis/build/collective/collective.mailchimp/src/collective/mailchimp/tests/test_portlet.py:41: PendingDeprecationWarning: Please use assertTrue instead.
  self.failUnless(IPortletDataProvider.providedBy(portlet.data))
    21/30 (70.0%)/home/travis/build/collective/collective.mailchimp/src/collective/mailchimp/tests/test_portlet.py:54: PendingDeprecationWarning: Please use assertTrue instead.
  self.failUnless(isinstance(mapping.values()[0], mailchimp.Assignment))
    22/30 (73.3%)/home/travis/build/collective/collective.mailchimp/src/collective/mailchimp/tests/test_portlet.py:62: PendingDeprecationWarning: Please use assertTrue instead.
  self.failUnless(isinstance(editview, mailchimp.EditForm))
    24/30 (80.0%)/home/travis/build/collective/collective.mailchimp/src/collective/mailchimp/tests/test_portlet.py:76: PendingDeprecationWarning: Please use assertTrue instead.
  self.failUnless(isinstance(renderer, mailchimp.Renderer))
    25/30 (83.3%)/home/travis/build/collective/collective.mailchimp/eggs/zope.deprecation-4.4.0-py2.7.egg/zope/deprecation/deprecation.py:88: DeprecationWarning: isDefaultPage is deprecated. Import from Products.CMFPlone.defaultpage instead
  name)
    28/30 (93.3%)/home/travis/build/collective/collective.mailchimp/src/collective/mailchimp/tests/test_setup.py:32: PendingDeprecationWarning: Please use assertTrue instead.
  self.failUnless(ICollectiveMailchimp in utils.registered_layers())
    29/30 (96.7%)
Failure in test test_mailchimp_css_available (collective.mailchimp.tests.test_setup.TestSetup)
Traceback (most recent call last):
  File "/opt/python/2.7.15/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/travis/build/collective/collective.mailchimp/src/collective/mailchimp/tests/test_setup.py", line 54, in test_mailchimp_css_available
    self.assertIn(key, resources.keys())
  File "/opt/python/2.7.15/lib/python2.7/unittest/case.py", line 804, in assertIn
    self.fail(self._formatMessage(msg, standardMsg))
  File "/opt/python/2.7.15/lib/python2.7/unittest/case.py", line 410, in fail
    raise self.failureException(msg)
AssertionError: 'resource-collective-mailchimp-stylesheets' not found in ['ace-mode-css', 'ace-mode-javascript', 'ace-mode-text', 'ace-theme-monokai', 'ace', 'backbone', 'backbone.paginator', 'backbone', 'bootstrap-alert', 'bootstrap-badges', 'bootstrap-basic', 'bootstrap-button-groups', 'bootstrap-buttons', 'bootstrap-collapse', 'bootstrap-dropdown', 'bootstrap-glyphicons', 'bootstrap-mixins', 'bootstrap-modal', 'bootstrap-progress-bars', 'bootstrap-tooltip', 'bootstrap-transition', 'bootstrap-variables', 'datatables.net-autofill-bs', 'datatables.net-autofill', 'datatables.net-bs', 'datatables.net-buttons-bs', 'datatables.net-buttons-colvis', 'datatables.net-buttons-flash', 'datatables.net-buttons-html5', 'datatables.net-buttons-print', 'datatables.net-buttons', 'datatables.net-colreorder', 'datatables.net-fixedcolumns', 'datatables.net-fixedheader', 'datatables.net-keytable', 'datatables.net-responsive-bs', 'datatables.net-responsive', 'datatables.net-rowreorder', 'datatables.net-scroller', 'datatables.net-select', 'datatables.net', 'dropzone', 'filemanager', 'jqtree-contextmenu', 'jqtree', 'jquery.browser', 'jquery', 'jquery.cookie', 'jquery', 'jquery.event.drag', 'jquery.event.drop', 'jquery', 'jquery.form', 'jquery', 'jquery.recurrenceinput', 'jquery.tmpl', 'jquery', 'less', 'logging', 'mockup-i18n', 'mockup-patterns-autotoc', 'mockup-patterns-backdrop', 'mockup-patterns-base', 'mockup-patterns-contentloader', 'mockup-patterns-cookietrigger', 'mockup-patterns-datatables', 'mockup-patterns-filemanager', 'mockup-patterns-formautofocus', 'mockup-patterns-formunloadalert', 'mockup-patterns-inlinevalidation', 'mockup-patterns-livesearch', 'mockup-patterns-markspeciallinks', 'mockup-patterns-modal', 'mockup-patterns-moment', 'mockup-patterns-navigationmarker', 'mockup-patterns-pickadate', 'mockup-patterns-preventdoublesubmit', 'mockup-patterns-querystring', 'mockup-patterns-recurrence', 'mockup-patterns-relateditems-upload', 'mockup-patterns-relateditems', 'mockup-patterns-resourceregistry', 'mockup-patterns-select2', 'mockup-patterns-sortable', 'mockup-patterns-structure', 'mockup-patterns-structureupdater', 'mockup-patterns-textareamimetypeselector', 'mockup-patterns-texteditor', 'mockup-patterns-thememapper', 'mockup-patterns-tinymce', 'mockup-patterns-toggle', 'mockup-patterns-tooltip', 'mockup-patterns-tree', 'mockup-patterns-upload', 'mockup-popover', 'mockup-router', 'mockup-ui', 'mockup-utils', 'moment', 'pat-base', 'pat-compat', 'pat-jquery-ext', 'pat-logger', 'pat-mockup-parser', 'pat-registry', 'pat-utils', 'picker', 'picker.date', 'picker', 'picker.time', 'picker', 'plone-logged-in', 'plone-patterns-portletmanager', 'plone-patterns-toolbar', 'plone', 'resource-plone-app-jquerytools-dateinput-js', 'resource-plone-app-jquerytools-js', 'resourceregistry', 'select2', 'text', 'thememapper', 'tinymce-advlist', 'tinymce-anchor', 'tinymce-autolink', 'tinymce-autoresize', 'tinymce-autosave', 'tinymce-bbcode', 'tinymce-charmap', 'tinymce-code', 'tinymce-colorpicker', 'tinymce-compat3x', 'tinymce-contextmenu', 'tinymce-default-styles', 'tinymce-directionality', 'tinymce-emoticons', 'tinymce-fullpage', 'tinymce-fullscreen', 'tinymce-hr', 'tinymce-image', 'tinymce-importcss', 'tinymce-insertdatetime', 'tinymce-legacyoutput', 'tinymce-link', 'tinymce-lists', 'tinymce-media', 'tinymce-modern-theme', 'tinymce-nonbreaking', 'tinymce-noneditable', 'tinymce-pagebreak', 'tinymce-paste', 'tinymce-preview', 'tinymce-print', 'tinymce-save', 'tinymce-searchreplace', 'tinymce-spellchecker', 'tinymce-tabfocus', 'tinymce-table', 'tinymce-template', 'tinymce-textcolor', 'tinymce-textpattern', 'tinymce-visualblocks', 'tinymce-visualchars', 'tinymce-wordcount', 'tinymce', 'translate', 'underscore']
```

Any takers? @pbauer any idea why things seems to have changed between Plone 5.1 and 5.2 here?

**Comment by erral** [comment]
I think that the CSS and JS thing is because of the removal of cssregistry.xml and jsregistry.xml support:

https://github.com/plone/Products.CMFPlone/issues/1742

So if we need those files, we would need to register them as resource registry resources and bundle I guess

**Comment by tisto** [comment]
@erral yep. Did you try the newsletter register/unregister forms on Plone 5.2? Do they still work or are they broken?

**Comment by erral** [comment]
I had to do some workarounds with the forms, but I thought that it was because I had to heavily customize the form, but now that I see that the tests fail in 5.2 that could be the reason... 

Let me check in a plain 5.2 and I'll report back and try to see what's going on in the failing test.

**Comment by erral** [comment]
The subscription view works, but has the portlet ever worked? I mean: the portlet has an option to restrict the list of available list to be subscribed to. But this list is never used and the form of the rendered portlet has always an empty list_id, so if you do not configure a default list_id in the product settings, you can never subscribe to anything.

**Comment by stevenhoog** [comment]
Hi everyone,

On Plone 5.2 (Python 3.6, zeocluster) I installed the collective.mailchimp module.
After the installation, 'Mailchimp' shows up in the controlpanel. So far, so good.

However, when I try to add MailChimp as a footer portlet, Only the portlet title appears. Any z3c form doesn’t show up. I ran the client in foreground mode, but I don’t see any warnings of a deprecated module that could be a dependency of the collective.mailchimp module.

Can someone point me into the direction of the actual problem? I'd be happy to contribute on solving the issue.

**Comment by tisto** [comment]
@stevenhoog I just tried c.mailchimp on Plone 5.2 with Python 3 and it seems to work. Though, you are correct that the mailchimp portlet does not work on the footer portlet manager. 

I just checked the basic functionality e.g. the newsletter registration page:

/newsletter

and adding a newsletter portlet:

![Welcome to Plone — Plone site](https://user-images.githubusercontent.com/632105/63860924-3f971500-c9aa-11e9-8ae1-5349cb135ca8.png)

That all works for me. So it could be that the newsletter portlet in the footer never worked in the first place. I don't see any tracebacks so I'd guess this might be a css problem.
